### PR TITLE
[SPARK-35594] Remove duplicate installations in build_and_test.yml

### PR DESCRIPTION
**Pull request by pingsutw**
### What changes were proposed in this pull request?

There are many duplicate installations of the python package in https://github.com/apache/spark/blob/a59063d5446a78d713c19a0d43f86c9f72ffd77d/.github/workflows/build_and_test.yml#L360-L386, so we could remove unnecessary installations.

### Why are the changes needed?

Remove unnecessary installations of the python package.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Pass the CIs

### CODE REVIEW

**Review by srowen — COMMENTED** [review]
These installs are for different purposes though - for example I don't think doc generation necessarily happens every time (I think?) But if linters don't need numpy it seems, then installing it for linters is not useful (and means you have to have the linter target run first). It might be an OK change after all but not sure it saves much.

**Comment by pingsutw** [comment]
@srowen Thanks for the review. 
In this GitHub Action workflow, those installations will always run before linter or doc generation. So I think we could remove it, and keep `build_and_test.yml` concise. 
And doc generation needs NumPy, so I move NumPy from linter to doc generation.

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -357,11 +357,7 @@ jobs:
         architecture: x64
     - name: Install Python linter dependencies
       run: |
-        # TODO(SPARK-32407): Sphinx 3.1+ does not correctly index nested classes.
```
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:360]: I would just keep them. I intentionally separated them for self-contained dependency list. 

Python linter also runs documentation build so we need the rest of dependencies

Comment (pingsutw) [reply .github/workflows/build_and_test.yml:360]: okay. Let's close this issue. Thanks for the review.

**Pull request closed without merging**